src/backend/cdb/cdbutil.c

```diff
@@ -261,7 +261,12 @@ readGpSegConfigFromCatalog(int *total_dbs)
 		attr = heap_getattr(gp_seg_config_tuple, Anum_gp_segment_configuration_warehouseid, RelationGetDescr(gp_seg_config_rel), &isNull);
 		Assert(!isNull);
 		warehouseid = DatumGetObjectId(attr);
-		if (OidIsValid(warehouseid) && warehouseid != GetCurrentWarehouseId())
+
+		/* content */
+		attr = heap_getattr(gp_seg_config_tuple, Anum_gp_segment_configuration_content, RelationGetDescr(gp_seg_config_rel), &isNull);
+		Assert(!isNull);
+
+		if (warehouseid != GetCurrentWarehouseId() && (OidIsValid(warehouseid) || DatumGetInt16(attr) != MASTER_CONTENT_ID))
```

Comment:
Is the logic correct here?
(OidIsValid(warehouseid) || DatumGetInt16(attr) != MASTER_CONTENT_ID)

Reply:
If we disable the warehouse feature, all warehouseid in gp_segment_configuration are 0(InvalidOid).

Then there are two cases:

1. If the current warehouse id is InvalidOid, all segments will be selected(default).
2. If the current warehouse id is valid, the master(warehouseid = 0, contentid = -1) and segments(warehouseid = current warehous id) will be selected.

It is the same if I write like this:
```
if (current_warehouseid == warehousid || conntent == MASTER_CONTENT_ID)
{
  // no continue
}
```
The downside of this approach is that it can make the number of code changes appear more.